File: Other/xkcdscrape.py
#####################################################################################
### The following function will go to the xkcd website, scrape the html, look for ###
### a particular part based on a div tag, snag it, and write it out to a file for ###
### subsequent text analysis.                                                     ###
#####################################################################################

### Inspiration from: 
### https://github.com/CabbagesAndKings/xkcd-Topics/blob/master/scripts/getTranscripts.sh
### See the R version in my repo.  Being new to Python I thought this would be a nice challenge.
### I make no claims as to its efficiency.

###############
### Preface ###
###############
import os
os.getcwd()

# change working directory to wherever you want the text files to be stored
os.chdir('C:/Users/mclark19/Desktop/CSR/Clients/Me/xkcdscrape/')

import time # for time comparisons if desired
import urllib2
import BeautifulSoup  

Soup = BeautifulSoup.BeautifulSoup 

###############
### Example ###
###############
### read in web contents
url = urllib2.urlopen("http://www.xkcd.com/1/")
xkcd1 = url.read()

### create a beautifulSoup object
out = Soup(xkcd1)

### Inspect in a standard html form
print(out.prettify())

### example of primary process: extract the div with id='transcript'
out.find('div', {'id' : 'transcript'}) 

######################################
### Main Process via Standard Loop ###
######################################

### Preliminaries
# number of xkcd comics
n = 1283

# test n
# n = 100

### set initial time
t0 = time.time() 

### Run the loop over all comics
for i in xrange(1, n+1):
    filename = 'xkcdpyScrape/xkcd' +  str(i) + '.txt'
    if i == 404:    # this deals with an xkcd joke for comic #404
        out = " "
    else:
        # read in the web contents
        url = urllib2.urlopen('http://www.xkcd.com/' + str(i))
        xkcd = url.read()
        
        # create a soup object
        xkcdsoup = Soup(xkcd)
        
        # scrape the div with id = 'transcript'
        out = xkcdsoup.find('div', {'id' : 'transcript'})
    
    # Write out the file
    f = open(filename, 'w+')
    f.write(str(out))
    f.close()

time.time() - t0  # roughly 20 seconds for 100, about 4.5 minutes for all

####################
### Parallelized ###
####################

##################################################################################
### Create scrapexkcd function that will do what's in the loop above given a   ###
### start and end point. See the loop above for details. Arguments include the ###
### start and end number specific to the comics one desires to download.       ###
##################################################################################

def scrapexkcd(start, end):
    import urllib2
    import BeautifulSoup 
    for i in xrange(start, end):
        filename = 'xkcdpyScrape/parallel/xkcd' +  str(i) + '.txt'
        if i == 404:
            out = " "
        else:
            url = urllib2.urlopen('http://www.xkcd.com/' + str(i))
            xkcd = url.read()
            xkcdsoup = BeautifulSoup.BeautifulSoup(xkcd)
            out = xkcdsoup.find('div', {'id' : 'transcript'})
        f = open(filename, 'w+')
        f.write(str(out))
        f.close()

start = 1
end = 1283 + 1 

# test end
end = 100

import pp # for parallelization

# set number of workers
ncpus = 3

# Create jobserver
job_server = pp.Server(ncpus=ncpus)

# Divide the task into subtasks
parts = 3
step = (end - start) / parts + 1

jobs = []

t0 = time.time()

for index in xrange(parts):
    # create start and endpoints for the function
    starti = start + index*step
    endi = min(start + (index+1)*step, end)
    # Submit a job which will scrape the site
    job_server.submit(scrapexkcd, (starti, endi))

#wait for jobs in all groups to finish 
job_server.wait()

time.time() - t0  # time since t0

# print stats
job_server.print_stats()

File: Other/xkcdscrape.R
#####################################################################################
### The following function will go to the xkcd website, scrape the html, look for ###
### a particular part based on a div tag, snag it, and write it out to a file for ###
### subsequent text analysis.                                                     ###
###                                                                               ###
### The function requires only one argument, number, indicating the number of     ###
### comic strips.                                                                 ###  
#####################################################################################

### Inspiration from: 
### https://github.com/CabbagesAndKings/xkcd-Topics/blob/master/scripts/getTranscripts.sh

xkcdscrape = function(number){
  # create url string
  url = paste0("http://xkcd.com/", number,'/') 
  
  # creates a single element list argument of unparsed html
  x = scrape(url, parse=F)
  x = as.character(x[[1]]) # not necessary, but simplifies things a bit
  
  # find the point in the text matching the regex
  transcript = regexpr('<div id="transcript" style="display: none">(.*?)</div>', x)
  
  # writes what is sent to console (via cat) to a xkcd#.txt file
  sink(paste0('xkcdRScrape/xkcd', number, '.txt')) # open connection to a file of the name formed by paste0; create the folder first.
  cat(unlist(regmatches(x, transcript))) # send to the console the point specified by transcript in the html text
  sink()  # close connection
}

library(parallel)
cl = makeCluster(3)      # nubmer of cores
clusterEvalQ(cl, library(scrapeR)) # load scrapeR package on the cores
clusterExport(cl, 'xkcdscrape') # export the function to the cores

n = 1283  # number of xkcd comics by late Oct 2013

# a standard non-parallel way to do it without an explicit loop and more straightforward/easier code
# sapply(1:n, xkcdscrape)  # each number 1:n is fed to the function

# the parallized version (timed via system.time function)
system.time({
  parSapply(cl, 1:n, xkcdscrape)
  }) 



### Example loop for comparison; slow
# Check website for number published
# n = 1283
# library(scrapeR)
# system.time({
# for (i in 1:n){
#   url = paste0("http://xkcd.com/", i,'/')
#   x = scrape(url, parse=F)
#   transcript = regexpr('<div id="transcript" style="display: none">(.*?)</div>', x[[1]])
#   sink(paste0('xkcdRScrape/xkcd', i))
#   cat(regmatches(x[[1]], transcript))
#   sink()
# }
# })